src/mc-text-search-str-encode.c

```diff
@@ -0,0 +1,250 @@
+    // If real_substrings and msize differ, we need to insert padding, so allocate one extra slot.
+    mc_substring_set_t *set = mc_substring_set_new(base_str,
+                                                   folded_len + 1,
+                                                   real_substrings == msize ? real_substrings : real_substrings + 1);
+    uint32_t idx = 0;
+    for (uint32_t i = lb; i < real_max_len + 1; i++) {
+        if (is_prefix) {
+            // [0, lb), [0, lb + 1), ..., [0, min(len, ub))
+            BSON_ASSERT(mc_substring_set_insert(set, 0, i, idx++, 1));
+        } else {
+            // [len - lb, len), [len - lb - 1, len), ..., [max(0, len - ub), len)
+            BSON_ASSERT(mc_substring_set_insert(set, folded_len - i, folded_len, idx++, 1));
+        }
+    }
+    if (msize != real_substrings) {
+        // Insert padding to get to msize
+        mc_substring_set_insert(set, 0, folded_len + 1, idx++, msize - real_substrings);
+    }
+    BSON_ASSERT(idx == set->n_indices);
+    return set;
+}
+
+static mc_substring_set_t *generate_suffix_tree(const char *base_str,
+                                                uint32_t folded_len,
+                                                uint32_t unfolded_len,
+                                                const mc_FLE2SuffixInsertSpec_t *spec) {
+    return generate_prefix_or_suffix_tree(base_str, folded_len, unfolded_len, spec->lb, spec->ub, false);
+}
+
+static mc_substring_set_t *generate_prefix_tree(const char *base_str,
+                                                uint32_t folded_len,
+                                                uint32_t unfolded_len,
+                                                const mc_FLE2PrefixInsertSpec_t *spec) {
+    return generate_prefix_or_suffix_tree(base_str, folded_len, unfolded_len, spec->lb, spec->ub, true);
+}
+
+static uint32_t calc_number_of_substrings(uint32_t strlen, uint32_t lb, uint32_t ub) {
+    // There are len - i + 1 substrings of length i in a length len string.
+    // Therefore, the total number of substrings with length between lb and ub
+    // is the sum of the integers inclusive between A = len - ub + 1 and B = len - lb + 1,
+    // A <= B. This has a closed form: (A + B)(B - A + 1)/2.
+    if (lb > strlen) {
+        return 0;
+    }
+    uint32_t largest_substr = MIN(strlen, ub);
+    uint32_t largest_substr_count = strlen - largest_substr + 1;
+    uint32_t smallest_substr_count = strlen - lb + 1;
+    return (largest_substr_count + smallest_substr_count) * (smallest_substr_count - largest_substr_count + 1) / 2;
+}
+
+static mc_substring_set_t *generate_substring_tree(const char *base_str,
+                                                   uint32_t folded_len,
+                                                   uint32_t unfolded_len,
+                                                   const mc_FLE2SubstringInsertSpec_t *spec) {
```

Comment:
This function also needs `beta` and `gamma` parameters which are the code-point lengths, not the byte length.

Reply:
done (folded codepoint len is inside the base string)